pkg/controller/trainingjob_controller.go

```diff
@@ -0,0 +1,258 @@
+package controller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	extv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	paddlev1 "github.com/paddlepaddle/edl/pkg/apis/paddlepaddle/v1"
+	paddleclient "github.com/paddlepaddle/edl/pkg/client/clientset/versioned"
+	paddlescheme "github.com/paddlepaddle/edl/pkg/client/clientset/versioned/scheme"
+	paddleinformers "github.com/paddlepaddle/edl/pkg/client/informers/externalversions"
+	paddlelisters "github.com/paddlepaddle/edl/pkg/client/listers/paddlepaddle/v1"
+	"github.com/paddlepaddle/edl/pkg/updater"
+)
+
+type TrainingJobController struct {
+	// KubeCli is a standard kubernetes clientset
+	KubeCli kubernetes.Interface
+	// ExtCli is the extension kubernetes clientset
+	ExtCli extclient.Interface
+	// PaddleCli is a clientset for our own API group
+	PaddleCli paddleclient.Interface
+
+	trainingjobLister paddlelisters.TrainingJobLister
+	trainingjobSynced cache.InformerSynced
+
+	// jobupdater is
+	jobupdater *sync.Map
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// New returns a trainingjob controller
+func New(
+	kubeCli kubernetes.Interface,
+	extCli extclient.Interface,
+	paddleCli paddleclient.Interface,
+	paddleInformers paddleinformers.SharedInformerFactory) *TrainingJobController {
+	trainingjobInformer := paddleInformers.Paddlepaddle().V1().TrainingJobs()
+	paddlescheme.AddToScheme(scheme.Scheme)
+
+	glog.V(4).Info("Creating trainingjob event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+	workqueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TrainingJob")
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "TrainingJobController"})
```

Comment:
Maybe we need some document or a sample controller to inform these concepts in a controller:

- informer
- event broadcaster
- workqueue
- recorder